### ISSUE TITLE
Training not using GPU desipte having tensorflow-gpu

### ISSUE THREAD

**Issue by YashRunwal** [matterport/Mask_RCNN#2336]
Hello guys,

I am doing custom object detection and instance segmentation using mask rcnn.

Hardware:
1. Nvidia Geforce 1050
2. Dell G3 15

Software:
I have installed all the packages listed in requirements.txt with following changes:-
1. tensorflow-gpu == 1.5.0
2. keras == 2.0.8

I have also installed Cuda Toolkit 9.0 and cudnn.

But the training is stuck on epoch 1/5

![image](https://user-images.githubusercontent.com/56769818/91024979-fe573680-e5f8-11ea-9205-0c7dd11888e9.png)

I have checked the tensorflow-gpu usage by using
tf.Session() and I can see that it can find my nvidia GPU: device 0. 
But it is not using the said gpu.

Can someone help me out?

Edit 1: 
I have read a bunch of issues related to this topic but I couldn't find any solution. 

I can see in my Task Manager that the GPU is not being utilised. 

I am not even sure if the training is started. 

I would really appreciate someone's help.

**Comment by burhr2** [comment]
Hi @YashRunwal  kindly check these issues https://github.com/matterport/Mask_RCNN/issues/2327#issuecomment-67893438 and https://github.com/matterport/Mask_RCNN/issues/2312. It has a pointer to an implementation that supports tensorflow2=> (with GPU).

**Comment by YashRunwal** [comment]
@Ab-Abdurrahman I have seen #2312 but I do not know which tensorflow (GPU) and keras version should I install. Can you help me out?

**Comment by burhr2** [comment]
@YashRunwal If you're using anaconda package. Kindly create a new env and install these package using `conda install package_name_below`

numpy
scipy
Pillow
cython
matplotlib
scikit-image
tensorflow-gpu
opencv-python
h5py
imgaug
IPython[all]

Then you can use the implementation from the issue https://github.com/matterport/Mask_RCNN/issues/2312

**Comment by YashRunwal** [comment]
Ahh okay. So it is not necessary to specify the version of tensorflow-gpu. And I see you have used conda install and not pip install. 

I will try this out and if it works I will close this issue or continue the thread. I hope this works because this will be my 3rd time of doing it all over again.

**Comment by nguyenkhaithinh** [comment]
Hello @YashRunwal, have you tried inserting these command line at the beginning of the script?
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"

**Comment by WDYIE** [comment]
Hey maybe you can only install the Tensorflow_gpu and uninstall cpu&amp;gpu one。



发自我的iPhone


------------------ Original ------------------

**Comment by YashRunwal** [comment]
@nguyenkhaithinh Yes I tried it but it still doesnt work. I have also tried with tensorflow 2 and cuda 10.1 but it still doesn't work. If I train the model on CPU and in Linux then it trains but if I do the same in Windows then the training doesn't even start! You have any idea why?

**Comment by YashRunwal** [comment]
@WDYIE I have tf-gpu==1.15
Then I tried with tf-gpu==1.5
Then I tried with tf-gpu==1.14

Still doesnt work.

**Comment by nguyenkhaithinh** [comment]
@YashRunwal well that's strange. I'm using tf 1.14.0 with keras 2.3.0 (CUDA10.0 and CuDNN 7.4.2) on both Windows 10 and LUbuntu 18.04; I installed both tf and keras via pip, not conda. I don't think Matterport Mask R-CNN is compatible with tf 2.0 for now.
Did you add CuDNN and CUDA folders to the system path? And did you mix tf, keras, CUDA and CuDNN versions?
Can you run this example to verify that keras and tf use GPU:
https://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

P/s: sorry for a mistake in the previous comment, it should be

import os
os.environ["CUDA_VISIBLE_DEVICES"] = **"0"** #Value in here is the 'PCI_BUS_ID' like this one https://github.com/matterport/Mask_RCNN/issues/109#issuecomment-348303194

**Comment by YashRunwal** [comment]
@nguyenkhaithinh So this is my detailed list of packages and their versions:

1. tensorflow-gpu = 1.15.0
2. keras = 2.0.8
3. numpy = 1.19.1
4. h5py = 2.10.1
5. opencv-python = 4.4.0.42
6. pycocotools = 2.0
7. python = 3.6.11

I have Cuda 9,10,11. Also the paths for Cuda 9 and 10 are added in the environment variables of Windows. The recent Nvidia driver is installed. I can also see that the tf detects nvidia and assigns it device:0 but it fails to use it in the script even after explicitly setting os.environment[]

Check the following image for CUDA and Nvidia versions:

![image](https://user-images.githubusercontent.com/56769818/91336574-6614a980-e7d2-11ea-9490-8b817f9c273d.png)

Why does it show in the image that the CUDA 11.0 is used? I mean yes, I have Cuda 11 too but it is not set in the environment variables

Also, how do we check the version of cudnn installed for 10.1 CUDA?

What is it that I am doing wrong? Are the versions mismatched?

**Comment by YashRunwal** [comment]
Also, @nguyenkhaithinh Do you have an example you can maybe share ?

**Comment by nguyenkhaithinh** [comment]
@YashRunwal I think it's because of version mismatch. About checking CuDNN version, as I follow this instruction https://askubuntu.com/questions/767269/how-can-i-install-cudnn-on-ubuntu-16-04, I check my CuDNN version by open the cudnn.h file (LUbuntu 18.04): /usr/local/cuda/include/cudnn.h. Scroll down a little bit and you will see something like these lines:

#define CUDNN_MAJOR 7
#define CUDNN_MINOR 4
#define CUDNN_PATCHLEVEL 1
==> My version is 7.4.1

I'm not quite sure which example you mean. If it's about installing tf-gpu, here is my configuration:
_tensorflow-gpu =1.14.0
_CUDA = 10.0
_CuDNN = 7.4.2 (windows 10)/ 7.4.1 (LUbuntu 18.04)
For pip, I think you should remove all current CUDA(s) and CuDNN to perform a clean installation. If you need different tf and CUDA versions, you can try installing them with Conda; it helps to solve version conflict.
You can refer to this link: https://www.tensorflow.org/install/source_windows#tested_build_configurations for more configuration.
After you re-run an example, could you please attach the output console to verify whether tf is using gpu or not, like this one (on windows):
________________________________________________________________________________________________________________________________
2020-08-27 11:31:53.441398: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this 
TensorFlow binary was not compiled to use: AVX2
**2020-08-27 11:31:53.444669: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library nvcuda.dll**  --> care for this line
2020-08-27 11:31:54.268204: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: 
name: GeForce GTX 1050 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.417
pciBusID: 0000:01:00.0
2020-08-27 11:31:54.268437: I tensorflow/stream_executor/platform/default/dlopen_checker_stub.cc:25] GPU libraries are statically linked, skip dlopen check.
2020-08-27 11:31:54.269286: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0
2020-08-27 11:31:56.686106: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:
2020-08-27 11:31:56.686269: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 
2020-08-27 11:31:56.686364: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N 
2020-08-27 11:31:56.687537: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 3002 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1050 Ti, pci bus id: 0000:01:00.0, compute capability: 6.1)
________________________________________________________________________________________________________________________________
P/s: I also use a laptop, MSI GF63 with GTX 1050Ti.

**Comment by YashRunwal** [comment]
@nguyenkhaithinh I will do this either today or tomorrow and share a screenshot if the error persists

**Comment by tortoiseme** [comment]
did anybody manage to resolve this issue? I have the same problem

**Comment by dhiren-hamal** [comment]
I solved this problem by removing the following lines of codes from Jupyter notebook:
```
os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = ""
```
The purpose of these lines of codes is for using multiple GPU for different notebooks simultaneously.
